[doc] Cryptographic Mailbox

swenson opened the pull request:
This adds sections describing the Cryptographic Mailbox system, its storage system, and related commands.

I cleaned up a little bit of the overview sections for what is new in the 1.1 and 2.0 runtime firmware to help keep things organized.

varuns-nvidia commented on runtime/README.md:
```diff
@@ -1,7 +1,22 @@
+
+## New in v2.0
+
+Caliptra 2.0 firmware supports passive mode (same as 1.x) and active (or susbystem) mode.
```
suggest replacing active with subsystem throughout the document to parallel the main spec. subsystem is misspelled here.

bluegate010 commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+| **Name**   | **Type** | **Description**                            |
+| ---------- | -------- | ------------------------------------------ |
+| chksum     | u32      |                                            |
+| usage      | u32      | Tag to specify how the data can be used    |
```
These tags are equal to the mailbox command codes?

bluegate010 commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+| chksum     | u32      |                                            |
+| usage      | u32      | Tag to specify how the data can be used    |
+|            |          | `CMGE` - AES-256-GCM key, must be 32 bytes |
+|            |          | `HMAC` - Can be used for HMAC, HKDF        |
```
Can I import a key of any size and use it as an HMAC key?

varuns-nvidia commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
 
 ### AUTHORIZE_AND_STASH
 
+The SoC uses this command to request authorization of its various SoC images. This command has the option to receive the image hash directly from SoC or from an external source (e.g., SHA Acc).
```
SHA Acc is deprecated in passive mode because the SHA is no longer expose on the AXI bus

jhand2 replied:
I think we still need to sort this out. This is a pretty annoying hw regression to deal in fw with since it impacts how the API surface works.

varuns-nvidia commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+
+### CM_HMAC
+
+Computes an HMAC with
```
unfinished sentence

bluegate010 commented on runtime/README.md:
```diff
@@ -1030,29 +1443,201 @@ The DPE `GET_CERTIFICATE_CHAIN` command shall return the following certificates:
+This scheme avoids having to store the IV while avoiding IV reuse problems.
+
+The use of GCM is to prevent the CMID from being modified outside of Caliptra.
+Although this concern is minor in the current design, where all of the keys are uniformly random bits, if the CMID ever uses additional bits to designate key usage (or other parameters), then GCM prevents an adversary from changing those paramters, changing an AES key to an ECDSA key, which has sensitive internal structure.
```
> if the CMID ever uses additional bits to designate key usage

`CM_IMPORT` allows the user to specify the usage for the key they are importing. Seems like we should reflect that usage constraint in the CMID.

swenson replied:
I forgot to change this in `CM_IMPORT`. I initially had the usage in the CMID, but I removed it for now since we don't have ECDSA and other sensitive keys. I am slightly inclined to keep this in a table in DCCM instead just to prevent this kind of attack completely.

Which, yes, would require augmenting the HKDF with a way to tag the output.

varuns-nvidia commented on runtime/README.md:
```diff
@@ -1030,29 +1443,201 @@ The DPE `GET_CERTIFICATE_CHAIN` command shall return the following certificates:
+
+CMIDs require no additional storage.
+
+Unecrypted CMIDs have the following structure:
```
typo unencrypted

jhand2 commented on runtime/README.md:
```diff
@@ -1,7 +1,22 @@
 
 This specification describes the Caliptra Runtime Firmware.
 
+## New in v1.1
```
nit: Might be nice to format these headings as a "Changelog" table

jhand2 commented on runtime/README.md:
```diff
@@ -58,20 +74,20 @@ Caliptra Runtime Firmware will share driver code with ROM and FMC where
-| `BAD_SIG`        | `0x4253_4947` ("BSIG") | Generic signature check failure (for crypto offload)
-| `BAD_IMAGE`      | `0x4249_4D47` ("BIMG") | Malformed input image
-| `BAD_CHKSUM`     | `0x4243_484B` ("BCHK") | Checksum check failed on input arguments
+| **Name**         | **Value**              | Description                                          |
```
nit: I would mildly prefer dropping these table formatting changes from this PR (or splitting them into a separate commit) to make it a bit easier to review the diff from a content perspective. I think they'll render in github the same either way.

But I understand if it's too much trouble to undo.

fdamato replied:
+1 on this

swenson replied:
Oops! I had reverted this in one of my revisions, but I must have accidentally turned back on table autoformatting. I'll undo these unrelated formatting changes.

jhand2 commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+| auth_req_result | u32      | Authorized: `0xDEADC0DE` |
+|                 |          | Denied: `0x21523F21`     |
+
+### SET_IMAGE_METADATA
```
Can you add a command description for this one? I'm not clear on what it's for.

jhand2 commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+|                |          | Value 2 = SHA2-384                |
+|                |          | Value 3 = SHA2-512                |
+| context size   | u32      | Size of the context               |
+| context        | u8[...]  |                                   |
```
Can the context be fixed-size? Since we use zerocopy for command serialization/deserialization, it's quite a bit simpler to handle if the interior of the request is fixed-size.

swenson replied:
For this particular one, yes, but I'd prefer to be flexible so that we don't have to change the format of the command if we decide to support other algorithms.

jhand2 commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+|                |          | Value 1 = SHA2-256                |
+|                |          | Value 2 = SHA2-384                |
+|                |          | Value 3 = SHA2-512                |
+| context size   | u32      | Size of the context               |
```
This API (and the hmac one) would probably be more clear using init/update/finish semantics.

fdamato replied:
I would agree to split, and keep a single command for single operations (no context would be required to be passed in this case)

swenson replied:
That's fair. I'll split them up.

jhand2 commented on runtime/README.md:
```diff
@@ -1,7 +1,22 @@
-# Caliptra Runtime Firmware v1.1
+# Caliptra Runtime Firmware v2.0
```
We should talk more about code size & ICCM size. I think we may have a hard time fitting all of this in 128KiB of SRAM.

swenson replied:
Agreed.

jhand2 commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+
+Command Code: `0x434D_5750` ("CMWP")
+
+`CM_WIPE` takes no input arguments.
```
nit: Command name is header is `CM_CLEAR`

jhand2 commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+| chksum      | u32      |                              |
+| fips_status | u32      | FIPS approved or an error    |
+
+### CM_CLEAR
```
I wonder if it would be helpful to report how much CM storage is in use in some status command.

swenson replied:
That's a good idea. I'll add that.

jhand2 commented on runtime/README.md:
```diff
@@ -1030,29 +1443,201 @@ The DPE `GET_CERTIFICATE_CHAIN` command shall return the following certificates:
+
+## Manifest-Based Image Authorization (1.2)
+
+Caliptra's goal is to enable integrators and vendors to meet standard security requirements for creating cryptographic identity and securely reporting measurements through DICE and DPE Certificate chains and Caliptra-owned private-public key pairs. In addition, Caliptra 1.0 provides an `ECDSA384_SIGNATURE_VERIFY` command to enable an SoC RoT to verify its own FW signatures so that it can develop  ...
```
nit: integrator and vendor generally refer to the same entity in Caliptra documentation.

jhand2 commented on runtime/README.md:
```diff
@@ -1030,29 +1443,201 @@ The DPE `GET_CERTIFICATE_CHAIN` command shall return the following certificates:
 current TCI Node to the root. Max of 32.
+
+
+## Manifest-Based Image Authorization (1.2)
```
nit: Since this is a pretty sizable PR, it would be helpful to split the CM changes and the Image Manifest documentation into separate PRs (or separate commits within this PR).

jhand2 commented on runtime/README.md:
```diff
@@ -1030,29 +1443,201 @@ The DPE `GET_CERTIFICATE_CHAIN` command shall return the following certificates:
+## Cryptographic Mailbox Commands (2.0)
+
+Cryptographic mailbox (CM) commands are a flexible set of mailbox commands that provide access to Caliptra's cryptographic cabilities.
+This is meant for key storage and use, supporting protocols like SPDM and OCP LOCK.
```
@bluegate010 IIRC, L.O.C.K. specifies specific commands that KMB (Caliptra) is supposed to implment. Is that correct?

If so, we may want to clarify that Caliptra L.O.C.K. firmware would implement the KMB APIs.

Also, because of the code size issue I mentioned above, we may have to think about whether L.O.C.K. integrations can get away with only implementing the KMB functions, or if they also need Caliptra Classic (TM) commands.

bluegate010 replied:
L.O.C.K. calls for more integrated flows, rather than purely being built in terms of crypto helpers. The integrated flows will need more firmware work, beyond what's offered in this interface. That being said, L.O.C.K. integrators _are_ likely going to want to leverage crypto helpers as part of their own firmware, so I'm not opposed to how the sentence is currently structured.

jhand2 commented on runtime/README.md:
```diff
@@ -1030,29 +1443,201 @@ The DPE `GET_CERTIFICATE_CHAIN` command shall return the following certificates:
+### Data Storage
+
+The CM system has its own dedicated storage in DCCM.
+This storage is protected with authenticated encryption using internal keys.
```
Are the encryption keys generated randomly on boot? Or derived deterministically?

jhand2 commented on runtime/README.md:
```diff
@@ -1030,29 +1443,201 @@ The DPE `GET_CERTIFICATE_CHAIN` command shall return the following certificates:
+
+### Storage Keys
+
+The key for the data referenced by a CMID is unique and derived via HKDF from the CDI<sub>RT</sub> with the info being the concatenation of `"CMID" || block offset || size`.
```
Could we generate a random number and put it in a keyvault slot at boot? Then whenever we need an AES key we can ask the KV slot of hmac the random seed + CMID metadata

Assuming we don't need these keys to live across cold reset.

bluegate010 replied:
Yes, that would be a preferred way of doing this, since durability is not required across cold boot.

swenson replied:
Using a random number in a keyvault slot was my first though, but that doesn't work for OCP LOCK, which needs persistence.

swenson replied:
Do we zeroize DCCM on power on? (It's SRAM, so I thought it would be persistent.)

But in any case, if LOCK can re-import the key after power on every time, then we can eliminate the persistence. (Though it begs the question of where is the LOCK key being stored if not Caliptra that it can be re-imported any time.)

bluegate010 commented on runtime/README.md:
```diff
@@ -1030,29 +1443,201 @@ The DPE `GET_CERTIFICATE_CHAIN` command shall return the following certificates:
+
+### CMIDs
+
+CMIDs require no additional storage.
```
Consider something like, "CMIDs offload storage of key metadata to mailbox callers. They therefore impose no additional storage within Caliptra Core."

bluegate010 commented on runtime/README.md:
```diff
@@ -1030,29 +1443,201 @@ The DPE `GET_CERTIFICATE_CHAIN` command shall return the following certificates:
+
+### Data Storage
+
+The CM system has its own dedicated storage in DCCM.
```
Does this mean that every key in CM storage is visible to Caliptra firmware? Seems like we should allow for keys that are not visible, such as shared secrets obtained from `CM_ECDH_FINISH`.

bluegate010 commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+| ----------- | -------- | -------------------------------- |
+| chksum      | u32      |                                  |
+| fips_status | u32      | FIPS approved or an error        |
+| output CMID | u8[32]   | Output CMID of the shared secret |
```
What kind of key is this, behind the CMID? Seems like it should be something that I can only pass to `CM_HKDF`.

swenson replied:
This will create a 384-bit shared secret. Theoretically it should only be used by HKDF, but other protocols could use the shared secret directly for AES.

bluegate010 replied:
We should document the secret size. And document that the AES commands will implicitly truncate the CMID key if it is too large.

bluegate010 commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+
+### CM_ECDH_GENERATE
+
+This computes the first half of an Elliptic Curve Diffie-Hellman exchange.
```
Probably worth specifying that this generates an ephemeral ECDH keypair, and the private portion is represented in `context`.

fdamato commented on runtime/README.md:
```diff
@@ -1030,29 +1443,201 @@ The DPE `GET_CERTIFICATE_CHAIN` command shall return the following certificates:
+
+Manifest-based image authorization is implemented via three mailbox commands: [`SET_AUTH_MANIFEST`](#set-auth-manifest), [`SET_IMAGE_METADATA`](#set-image-metadata), and [`AUTHORIZE_AND_STASH`](#authorize-and-stash).
+
+### Caliptra-Endorsed Aggregated Secure Boot
```
wondering if this should be called "Aggregated Measured Boot"

fdamato replied:
The confusion here in calling it "Secure Boot" is that when Caliptra is Passive, it has no "power" to enforce what it is running

fdamato commented on runtime/README.md:
```diff
@@ -1030,29 +1443,201 @@ The DPE `GET_CERTIFICATE_CHAIN` command shall return the following certificates:
+
+A local verifier provides an authentication of SoC FW by matching SoC FW measurements with measurements from the Caliptra measurement manifest. In this case, the SoC RoT still has its own FW public-key chain that is verified by the SoC RoT, but in addition the SoC RoT introduces the Caliptra Measurement Manifest, which is endorsed by the Caliptra FW key pair. Caliptra provides approval or disappr ...
+
+The Caliptra-Endorsed Local Verify could be required by the owner only or both the vendor and the owner.
```
typo: verifier ?

fdamato commented on runtime/README.md:
```diff
@@ -1030,29 +1443,201 @@ The DPE `GET_CERTIFICATE_CHAIN` command shall return the following certificates:
+
+### Caliptra-Endorsed Local Verifier
+
+A local verifier provides an authentication of SoC FW by matching SoC FW measurements with measurements from the Caliptra measurement manifest. In this case, the SoC RoT still has its own FW public-key chain that is verified by the SoC RoT, but in addition the SoC RoT introduces the Caliptra Measurement Manifest, which is endorsed by the Caliptra FW key pair. Caliptra provides approval or disappr ...
```
Should the manifest be endorsed by the Owner Key Pair ?

swenson commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+| last length     | u32      |                 |
+| last counter    | u[16]    |                 |
+| last GHASH      | u[16]    |                 |
+| last ciphertext | u[16]    |                 |
```
The last counter is the last plaintext.

But honestly this is just an implementation detail and we can probably just delete it.

swenson commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+| fips_status  | u32      | FIPS approved or an error                 |
+| context size | u32      | Can be 0 if this is the final message     |
+| context      | u8[...]  |                                           |
+| hash size    | u32      | Can be 0 if this is not the final message |
```
I don't love having magic values, even if all 0s. Though it is probably safe here.

swenson commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+|                |          | 1 = SHA2-256       |
+|                |          | 2 = SHA2-384       |
+|                |          | 3 = SHA2-512       |
+| data size      | u32      |                    |
```
Good point. I'll add that in.

swenson commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+| last length     | u32      |                 |
+| last counter    | u[16]    |                 |
+| last GHASH      | u[16]    |                 |
+| last ciphertext | u[16]    |                 |
```
Just the last 16 bytes of the ciphertext, which the next invocation will use.

swenson commented on runtime/README.md:
```diff
@@ -692,107 +708,504 @@ Command Code: `0x434B_4558` ("CKEX")
+| **Name**       | **Type** | **Description**    |
+| -------------- | -------- | ------------------ |
+| chksum         | u32      |                    |
+| Key CMID       | u8[32]   | CMID of key to use |
```
Almost every use of the CMID is slightly different (e.g., IKM, key), so I thought it best to spell it out each time.

swenson commented:
I believe I have addressed the comments relevant to the cryptographic mailbox commands. I'm creating a separate PR for the manifest-based image auth doc updates, as suggested.

PTAL.

bluegate010 commented on runtime/README.md:
```diff
@@ -692,107 +787,631 @@ Command Code: `0x434B_4558` ("CKEX")
+| context size       | u32      | size of context                       |
+| context            | u8[...]  | Used as the input to `CM_ECDH_FINISH` |
+| exchange data size | u32      | size of data to send in the exchange  |
+| exchange data      | u8[...]  | i.e., the public point                |
```
Seems like we should document the format of this point? Is it 0x04 || x || y?

swenson replied:
I'm not even sure that `0x04` is necessary. I don't see it in the TLS exchange data at least. I'll document that.

bluegate010 commented on runtime/README.md:
```diff
@@ -692,107 +787,631 @@ Command Code: `0x434B_4558` ("CKEX")
+| context size                | u32      | size of context                                          |
+| context                     | u8[...]  | This MUST come from the output of the `CM_ECDH_GENERATE` |
+| incoming exchange data size | u32      |                                                          |
+| incoming exchange data      | u8[...]  | the other side's public point                            |
```
Ditto.

bluegate010 commented on runtime/README.md:
```diff
@@ -692,107 +787,631 @@ Command Code: `0x434B_4558` ("CKEX")
+
+Command Code: `0x434D_4B50` ("CMKP")
+
+*Table: `CM_HKDF` input arguments*
```
s/CM_HKDF/CM_HKDF_EXPAND/?

bluegate010 commented on runtime/README.md:
```diff
@@ -692,107 +787,631 @@ Command Code: `0x434B_4558` ("CKEX")
+| ---------- | -------- | ------------------------------------------ |
+| chksum     | u32      |                                            |
+| usage      | u32      | Tag to specify how the data can be used    |
+|            |          | 0 - Can be used for HMAC, HKDF, AES        |
```
I feel like we could be more prescriptive. Something like,

```
0 - HMAC-SHA2-256
1 - HMAC-SHA2-384
2 - HMAC-SHA2-512
3 - HKDF-SHA2-256
4 - HKDF-SHA2-384
5 - HKDF-SHA2-512
6 - AES-256-GCM
```
Then each key can only be used for a specific purpose, promoting good hygiene. CMIDs returned from `CM_ECDH_FINISH` could be implicitly tagged with type `HKDF-SHA2-384`. Callers could explicitly pass a value from this list as an arg to `CM_HKDF_EXPAND` (which could potentially replace the "Output length" arg).

swenson replied:
I'm about 50% agreed, except that it is really hard to differentiate between keys for HMAC and HKDF, and TLS isn't required to use HKDF (it's only a `SHOULD`).

bluegate010 replied:
A TLS implementation would know whether a given key is about to be used for HMAC or HKDF, so the caller should be able to specify at each step. To support either case for a key that comes out of `CM_HKDF_EXPAND`, we could allow the caller to specify the purpose as an input arg to that function.

swenson commented on runtime/README.md:
```diff
@@ -58,7 +73,84 @@ Caliptra Runtime Firmware will share driver code with ROM and FMC where
+
+### CMIDs
+
+CMIDs offload storage of key metadata to mailbox callers. They therefore impose no additional storage within Caliptra Core.
```
I thought about this, and the main problem is the size of systems like ML-KEM, which we will almost certainly be supporting in 2.1+.

bluegate010 replied:
ML-KEM private keys are stored as seeds. Not sure why that's motivating consuming limited Caliptra memory, vs the caller's.

swenson commented on runtime/README.md:
```diff
@@ -692,107 +787,631 @@ Command Code: `0x434B_4558` ("CKEX")
+| fips_status    | u32      | FIPS approved or an error                 |
+| context size   | u32      | SHALL be 0 if this is the final message   |
+| context        | u8[...]  |                                           |
+| cipertext size | u32      | MAY be 0 if this is not the final message |
```
It depends on the size of the input plaintext. If they send less than 16 bytes or our buffer is not full, we may not have output.

But we aren't going to buffer the entire output until the final call as that would not allow us to do large requests.

swenson commented on runtime/README.md:
```diff
@@ -692,107 +787,631 @@ Command Code: `0x434B_4558` ("CKEX")
+| chksum         | u32      |                                      |
+| context size   | u32      | Size of the context.                 |
+| context        | u8[...]  | From `CM_SHA_INIT` / `CM_SHA_UPDATE` |
+| data size      | u32      | May be 0                             |
```
We could. But then we get back to where we before. Because if we have an empty data, then we could really skip to the update (since the init wouldn't actually be doing anything other than calculating the starting value). And then we might as well combine the final and update steps. :)

swenson commented on runtime/README.md:
```diff
@@ -692,107 +787,631 @@ Command Code: `0x434B_4558` ("CKEX")
+| context size   | u32      | SHALL be 0 if this is the final message   |
+| context        | u8[...]  |                                           |
+| cipertext size | u32      | MAY be 0 if this is not the final message |
+| ciphertext     | u8[...]  |                                           |
```
Therein lies danger. :) FIPS generally doesn't certify implementations that have user-generated IVs. I think it is better to be prescriptive and say the IV is always random. We could revisit this if OCP Lock or another use case needs, say, a deterministic IV.

bluegate010 replied:
`CM_AES_GCM_DECRYPT` requires the caller to pass an IV. Where is the IV supposed to come from, if it's not returned from `CM_AES_GCM_ENCRYPT`?

swenson commented on runtime/README.md:
```diff
@@ -1,7 +1,22 @@
-# Caliptra Runtime Firmware v1.1
+# Caliptra Runtime Firmware v2.0
```
Done

swenson commented:
Apologies for merging a little early, but we wanted to have this in before the deadline. I'm happy to do further revisions and have more discussions.

bluegate010 commented on runtime/README.md:
```diff
@@ -692,107 +791,628 @@ Command Code: `0x434B_4558` ("CKEX")
+
+The context must be passed from the `CM_ECDH_GENERATE` command.
+
+The incoming exchange data MUST be the concatenation of the x- and y- coordinates of the other side's public point.
```
How are the coordinates encoded? Big-endian?

swenson replied:
Yes, big-endian.

raghuncstate commented on runtime/README.md:
```diff
@@ -58,7 +77,84 @@ Caliptra Runtime Firmware will share driver code with ROM and FMC where
+
+The contexts contain the internal structures necessary to resume operations to support data that may exceed the size of a single mailbox command.
+
+These contexts are intended to be opaque to the user, and SHALL be encrypted and authenticated if they contain sensitive internal data.
```
where is it stored? and why not replay protected?

raghuncstate commented on runtime/README.md:
```diff
@@ -58,7 +77,84 @@ Caliptra Runtime Firmware will share driver code with ROM and FMC where
+
+The contexts contain the internal structures necessary to resume operations to support data that may exceed the size of a single mailbox command.
+
+These contexts are intended to be opaque to the user, and SHALL be encrypted and authenticated if they contain sensitive internal data.
```
what threat is this intended to solve/mitigate? it is defense in depth mostly?
Asked another way, what happens if you leave it unencrypted? presumably, a runtime attack on FW has access to the key (indirectly?) used to encrypt? or is that not hte case?

swenson replied:
The context might have intermediate state that is sensitive. For example, intermediate GHASH computations in AES could be used to reveal the additional authenticated data, which may or may not be sensitive.

The authenticated encryption is more defense in depth.

raghuncstate replied:
thx, agree on the intermediate state being sensitive, but was wondering HOW you envision an attacker getting access to it, mostly to try and answer the question: if the attacker can get access to context data, can the attacker also not get access to the key?
What makes getting access to the key (or its handle) more or less difficult than having access to the context data?

Note, not asking for change, trying to better understand thought process behind decision to do so.

raghuncstate replied:
i think i may have misread/misunderstood. I thought contexts are stored in the data storage, but i see that you provide it as output in commands below, so the question above may be moot, but would appreciate confirmation that data storage below, is for the KEYS, not contexts :)
Do we care about replay protection of the contexts?

swenson replied:
Contexts could be stored insecurely in the SoC, so they are more vulnerable to being intercepted.

In this version, the keys are only accessible through handles, but the keys themselves are stored only in DCCM in Caliptra. (This may change so that the keys are also returned to the SoC in encrypted form to minimize storage requirements in Caliptra.)

Contexts are also encrypted and authenticated. For many of the commands, I don't think we're too worried about replay attacks (e.g., SHA). But, I think you're right that some commands could use replay attack protection. For instance, with AES-GCM, if you have a context and an associated ciphertext block, you could replay the context with a new plaintext and use that to recover the other plaintext.

Theoretically this is all over a trusted channel, so these attacks may not be part of the threat model, but we should be able to apply some internal counters in DCCM to help with this. I've filed an issue to address this: https://github.com/chipsalliance/caliptra-sw/issues/1761

raghuncstate commented on runtime/README.md:
```diff
@@ -58,7 +77,84 @@ Caliptra Runtime Firmware will share driver code with ROM and FMC where
+
+The IV is the concatenation of `block offset || size`.
+
+This scheme avoids having to store the IV while avoiding IV reuse problems.
```
can you elaborate on why IV reuse is not a problem? if you store the different data twice at the same offset with same size, would you not generate the same key and same IV again to encrypt the new data?

swenson replied:
I noticed that after I wrote it. It's definitely true (if the data being stored is the same length).

I'm debating a few different ways to fix this, but essentially involve storing the IV or a part of the IV either in DCCM or with the CMID.

bluegate010 replied:
> if you store the different data twice at the same offset with same size, would you not generate the same key and same IV again to encrypt the new data?

I don't think you would, right? Because the CMID does not hold any data related to the key itself, just the offset and size. So if you store two different keys with the same size at the same offset, the same key will encrypt the same CMID.

raghuncstate replied:
okay, maybe i misunderstood, what this is encrypting. I thought this is a key, being derived, to encrypt context data. The key being derived, look like a deterministic key HKDF(CDI_RT, "CMID||block offset||size"), and that key would AES GCM the context data (which can be different data bits).
I did not think this is the key used to encrypt the CMID itself, given to the user.

swenson replied:
The scheme here was `K = HKDF(CDI_RT, block offset || size)`, which would then be used to encrypt that portion of DCCM that holds the key associated with the CMID.

But I think we're going to move away from storing keys in DCCM shortly, so this shouldn't matter. Though we will still have to be very careful about IV reuse.

raghuncstate commented on runtime/README.md:
```diff
@@ -692,107 +791,628 @@ Command Code: `0x434B_4558` ("CKEX")
+| chksum         | u32      |                                      |
+| context size   | u32      | Size of the context.                 |
+| context        | u8[...]  | From `CM_SHA_INIT` / `CM_SHA_UPDATE` |
+| data size      | u32      |                                      |
```
do we care that data size field may become an unaligned access depending on context size?

bluegate010 replied:
I think we could define a fixed-size context struct, sized to hold the largest context data possible, and make sure it's aligned properly.

swenson replied:
I worry about a fixed-size struct, as it means that if we define a new command or want to redefine the context for one command, we break compatibility with all commands that use contexts. With the variable-sized contexts, we wouldn't need to break compatibility if we needed to add a field to any context.

But we can stipulate that all contexts have sizes that are multiples of 32 bits.
